library/ceph_key.py

```diff
@@ -0,0 +1,422 @@
+    '''
+
+    if module:
+        module.fail_json(msg=message, rc=1)
+    else:
+        raise(Exception(message))
+
+
+def key_mode(file_path):
+    '''
+    Change mode file for a CephX key
+    Problem, how do to do on containerized deployment?
+    '''
+    os.chmod(file_path)
+
+
+def generate_secret():
+    '''
+    Generate a CephX secret
+    '''
+
+    key = os.urandom(16)
+    header = struct.pack('<hiih', 1, int(time.time()), 0, len(key))
+    secret = base64.b64encode(header + key)
+
+    return secret
+
+
+def generate_caps(cmd, type, caps):
+    '''
+    Generate CephX capabilities list
+    '''
+    cmd = cmd
+
+    for k, v in caps.iteritems():
+        if type == "ceph-authtool":
+            cmd.extend(["--cap"])
+
+        cmd.extend([k, v])
+
+    return cmd
+
+
+def generate_ceph_cmd(cluster, args, containerized):
+    '''
+    Generate 'ceph' command line to execute
+    '''
+
+    cmd = []
+
+    base_cmd = [
+        'ceph',
+        '--cluster',
+        cluster,
+        'auth',
+    ]
+
+    cmd.extend(base_cmd + args)
+
+    if containerized:
+        cmd = containerized.split() + cmd
+
+    return cmd
+
+
+def generate_ceph_authtool_cmd(cluster, name, secret, caps, containerized):
+    '''
+    Generate 'ceph-authtool' command line to execute
+    '''
+
+    file_destination = "/etc/ceph/" + cluster + "." + name + ".keyring"
+
+    cmd = [
+        'ceph-authtool',
+        '--create-keyring',
+        file_destination,
+        '--name',
+        name,
+        '--add-key',
+        secret,
+    ]
+
+    cmd = generate_caps(cmd, "ceph-authtool", caps)
+
+    if containerized:
+        cmd = containerized.split() + cmd
+
+    return cmd
+
+
+def create_key(module, result, cluster, name, secret, caps, containerized):
+    '''
+    Create a CephX key
+    '''
+
+    file_path = "/etc/ceph/" + cluster + "." + name + ".keyring"
+
+    args = [
+        "import",
+        "-i",
+        file_path,
+    ]
+    cmd_list = []
+
+    if secret is None:
+        secret = generate_secret()
+
+    cmd_list.append(generate_ceph_authtool_cmd(cluster, name, secret, caps, containerized))
+    cmd_list.append(generate_ceph_cmd(cluster, args, containerized))
+
+    return cmd_list
+
+
+def update_key(cluster, name, caps, containerized):
+    '''
+    Update a CephX key's capabilities
+    '''
+
+    cmd_list = []
+
+    args = [
+        "caps",
+        name,
+    ]
+
+    args = generate_caps(args, "ceph", caps)
+    cmd_list.append(generate_ceph_cmd(cluster, args, containerized))
+
+    return cmd_list
+
+
+def delete_key(cluster, name, containerized):
+    '''
+    Delete a CephX key
+    '''
+
+    cmd_list = []
+
+    args = [
+        "del",
+        name,
+    ]
+
+    cmd_list.append(generate_ceph_cmd(cluster, args, containerized))
+
+    return cmd_list
+
+
+def info_key(cluster, name, containerized):
+    '''
+    Get information about a CephX key
+    '''
+
+    cmd_list = []
+
+    args = [
+        "get",
```

Comment:
this is just nitpicking, but seems like you are mixing double quotes with single quotes. I don't mind either, but would be nice to just stick with either or.

Reply:
good catch thanks